File: src/main/java/com/mcmoddev/basesciences/common/tiles/TileEntityMultiblock.java
package com.mcmoddev.basesciences.common.tiles;

import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;

import javax.annotation.Nonnull;

public class TileEntityMultiblock extends TileEntity {

    private boolean hasControllerBlock = false;
    private boolean isControllerBlock = false;
    private BlockPos controllerBlockPos = BlockPos.ORIGIN;

    public TileEntityMultiblock() {
        controllerBlockPos = BlockPos.ORIGIN;
    }

    public boolean getHasControllerBlock() {
        return hasControllerBlock;
    }

    public void setHasControllerBlock(boolean hasControllerBlock) {
        this.hasControllerBlock = hasControllerBlock;
    }

    public boolean isControllerBlock() {
        return this.isControllerBlock;
    }

    public void setControllerBlock(boolean isControllerBlock) {
        this.isControllerBlock = isControllerBlock;
    }

    public BlockPos getControllerBlockPos() {
        return this.controllerBlockPos;
    }

    public void setControllerBlockPos(BlockPos controllerBlockPos) {
        this.controllerBlockPos = controllerBlockPos;
    }


    @Nonnull
    @Override
    public NBTTagCompound writeToNBT(@Nonnull NBTTagCompound compound) {
        super.writeToNBT(compound);
        compound.setBoolean("hasControllerBlock", hasControllerBlock);
        compound.setBoolean("isControllerBlock", isControllerBlock);
        compound.setInteger("controllerBlockXPos", controllerBlockPos.getX());
        compound.setInteger("controllerBlockYPos", controllerBlockPos.getY());
        compound.setInteger("controllerBlockZPos", controllerBlockPos.getZ());
        return compound;
    }

    @Override
    public void readFromNBT(@Nonnull NBTTagCompound compound) {
        super.readFromNBT(compound);
        hasControllerBlock = compound.getBoolean("hasControllerBlock");
        isControllerBlock = compound.getBoolean("isControllerBlock");
        controllerBlockPos = new BlockPos(compound.getInteger("controllerBlockXPos"),
                compound.getInteger("controllerBlockYPos"), compound.getInteger("controllerBlockZPos"));
    }
}

File: src/main/kotlin/com/mcmoddev/basesciences/tile/TileEntityMultiblock.kt
package com.mcmoddev.basesciences.tile

import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.math.BlockPos

open class TileEntityMultiblock: TileEntity() {
    var hasMaster = false
    var isMaster = false
    var masterPos: BlockPos = BlockPos.ORIGIN

    override fun writeToNBT(compound: NBTTagCompound?): NBTTagCompound {
        super.writeToNBT(compound!!)
        compound.setBoolean("hasMaster", hasMaster)
        compound.setBoolean("isMaster", isMaster)
        compound.setInteger("masterX", masterPos.x)
        compound.setInteger("masterY", masterPos.y)
        compound.setInteger("masterZ", masterPos.z)
        return compound
    }

    override fun readFromNBT(compound: NBTTagCompound?) {
        super.readFromNBT(compound!!)
        hasMaster = compound.getBoolean("hasMaster")
        isMaster = compound.getBoolean("isMaster")
        masterPos = BlockPos(compound.getInteger("masterX"), compound.getInteger("masterY"), compound.getInteger("masterZ"))
    }
}